# Введение в NumPy

## Основная информация и создание

Главный объект NumPy - это однородный многомерный массив.

однородный - один тип

многомерный - N измерения, от матрицы с 2 измерениями до ...

Чаще всего это одномерная последовательность или двумерная таблица, заполненная элементами одного типа, как правило числами, которые проиндексированы кортежем положительных целых чисел.

- ***``Numpy`` - это библиотека Python для вычислительно эффективных операций с многомерными массивами, предназначенная в основном для научных вычислений.***


- ***Пакет ``Numpy`` предоставляет $n$-мерные однородные массивы (все элементы одного типа) в них нельзя вставить или удалить элемент в произвольном месте. В ``Numpy`` реализовано много операций над массивами в целом.***

### Создание


In [1]:
import numpy as np

In [2]:
# создание из списка  np.array()

import numpy as np

a = np.array([1, 2, 3, 4])
b = np.array((5, 6, 7, 8))
c = np.array([[1, 2, 3], [4, 5, 6]])
print(a)  # [1 2 3 4]
print(c)  # [[1 2 3]
          #  [4 5 6]]


[1 2 3 4]
[[1 2 3]
 [4 5 6]]


In [3]:
# Создание массивов с заранее заданными значениями

zeros_arr = np.zeros((3, 3))
ones_arr = np.ones(5)
full_arr = np.full((2, 4), 7)
print(zeros_arr)
print(ones_arr)
print(full_arr)



[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]
[1. 1. 1. 1. 1.]
[[7 7 7 7]
 [7 7 7 7]]


In [4]:
# Создание массивов с использованием функций генерации

arr1 = np.arange(0, 10, 2)  # [0 2 4 6 8]   - cоздает массив с шагом
arr2 = np.linspace(0, 1, 5) # [0.   0.25 0.5  0.75 1. ] - создает массив из num равномерно распределенных чисел между start и stop


In [5]:
# Создание массивов с помощью функции по координатам элементов

def rule(i, j):
    return i + j

arr = np.fromfunction(rule, (3, 3))
print(arr)
# [[0. 1. 2.]
#  [1. 2. 3.]
#  [2. 3. 4.]]


[[0. 1. 2.]
 [1. 2. 3.]
 [2. 3. 4.]]


In [6]:
# Копирование существующих массивов
origin = np.array([1, 2, 3])
copy_arr = origin.copy()


In [7]:
# создание матрицы единиц размерностью матрица "а"
b = np.ones_like(c)
b, b.shape

(array([[1, 1, 1],
        [1, 1, 1]]),
 (2, 3))

In [8]:
# создание пустой матрицы (случайные элементы)
c = np.empty((3,6), dtype='int')
c

array([[                  0,                   0,                   0,
        4607182418800017408, 4607182418800017408, 4607182418800017408],
       [4611686018427387904, 4611686018427387904, 4611686018427387904,
                          0, 4607182418800017408, 4611686018427387904],
       [                  0, 4607182418800017408, 4611686018427387904,
                          0, 4607182418800017408, 4611686018427387904]])

### list comprehension 

list comprehension — мощный Python-инструмент для создания списков; с NumPy лучше предпочитать векторные операции для производительности. 

In [9]:
# List comprehension с массивами NumPy

import numpy as np

mass = np.array([1, 2, 3, 4, 5])

# Создаём список квадратов элементов mass
list_squares = [x**2 for x in mass]
print(list_squares)  # [1, 4, 9, 16, 25]

# Создаём массив из list comprehension (обратно в numpy.ndarray)
arr_squares = np.array([x**2 for x in mass])
print(arr_squares)   # [ 1  4  9 16 25]


[np.int64(1), np.int64(4), np.int64(9), np.int64(16), np.int64(25)]
[ 1  4  9 16 25]


In [10]:
# Векторизированные операции — лучше чем list comprehension

import numpy as np

mass = np.array([1, 2, 3, 4, 5])

# Возводим в квадрат все элементы сразу (векторное умножение)
arr_squares = mass ** 2
print(arr_squares)  # [ 1  4  9 16 25]


[ 1  4  9 16 25]


In [11]:
# Создание массивов с помощью list comprehension внутри np.array()

import numpy as np

# Создадим массив, где элементы от 0 до 9 возведены в квадрат,
# но только для чётных чисел, остальные заменены на -1
arr_custom = np.array([x**2 if x % 2 == 0 else -1 for x in range(10)])
print(arr_custom)
# [-1  -1   4  -1  16  -1  36  -1  64  -1]


[ 0 -1  4 -1 16 -1 36 -1 64 -1]


### Основные операции

In [12]:
lst = [11, 22, 33, 44, 55, 66, 77, 88, 99]
mass = np.array(lst, dtype=np.int32) # вектор
print(mass)
print(len(mass))

[11 22 33 44 55 66 77 88 99]
9


Теперь у нас есть одномерный массив, т.е. у него всего одна ось вдоль которой происходит индексирование его элементов.

![image](https://raw.githubusercontent.com/dm-fedorov/numpy_basic/master/pic/1nd_array.jpg)

In [13]:
# Генерация списка из элементов вектора

new_lst = [0]*9 # генерируем list из 0 len = 9
indices = 0
for val in mass:
  # foreach
  new_lst[indices] = val
  indices +=1
print(new_lst)

[np.int32(11), np.int32(22), np.int32(33), np.int32(44), np.int32(55), np.int32(66), np.int32(77), np.int32(88), np.int32(99)]


In [14]:
# создание нового вектора из старого по индексам
new_mass = mass[[7, 0, 3, 3, 3, 0, 7]] # двойные скобки позволяют взять элементы по индексу
new_mass # новый из массива выше

array([88, 11, 44, 44, 44, 11, 88], dtype=int32)

In [15]:
new_lst[[7, 0, 3, 3, 3, 0, 7]] # нельзя так!

TypeError: list indices must be integers or slices, not list

Вместо одного индекса, указан целый список индексов. А вот еще любопытный пример, теперь вместо индекса укажем логическое выражение:

In [16]:
# Взять элементы вектора по условию

mass[mass > 50] # в скобках условие

array([55, 66, 77, 88, 99], dtype=int32)

In [17]:
# двойное условие
mass[(mass > 50) & (mass < 80)] 

array([55, 66, 77], dtype=int32)

In [18]:
new_lst[new_lst > 50] # Ошибка  - нужно, чтобы new_lst был NumPy массивом, а не списком.


TypeError: '>' not supported between instances of 'list' and 'int'

Цель этих двух примеров - не устраивать головоломку, а продемонстрировать расширенные возможности индексирования массивов NumPy.

Что еще интересного можно продемонстрировать? Векторизованные вычисления:

In [19]:
2*mass + 10

array([ 32,  54,  76,  98, 120, 142, 164, 186, 208], dtype=int32)

In [20]:
np.sin(mass)**2 + np.cos(mass)**2 # np - numpy сокращение

array([1., 1., 1., 1., 1., 1., 1., 1., 1.])

Векторизованные - означает, что все арифметические операции и математические функции выполняются сразу над всеми элементами массивов. А это в свою очередь означает, что нет никакой необходимости выполнять вычисления в цикле.

Давайте перейдем к двумерным массивам:

In [21]:
gen_X = np.arange(12) # не включая последнее число, с 0
gen_X

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [22]:
matrix_gen_X = gen_X.reshape(3, 4)
matrix_gen_X

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

Сейчас мы создали массив с помощью функции `np.arange()`, которая во многом аналогична функции `range()` языка Python.

Затем, мы изменили форму массива с помощью метода `reshape()`, т.е. на самом деле создать этот массив мы могли бы и одной командой:

In [ ]:
a = np.arange(12).reshape(3, 4)
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

Визуально, данный массив выглядит следующим образом:

![image](https://raw.githubusercontent.com/dm-fedorov/numpy_basic/master/pic/2nd_array.jpg)

Глядя на картинку, становится понятно, что первая ось (и индекс соответственно) - это строки, вторая ось - это столбцы. Т.е. получить элемент 9 можно простой командой:

In [ ]:
a[2][1]    #  равносильно команде a[2, 1]

np.int64(9)

Снова можно подумать, что ничего нового - все как в стандартном Python.
Да, так и есть, и, это круто!

Еще круто, то что NumPy добавляет к удобному и привычному синтаксису Python, весьма удобные трюки, например - *транслирование массивов*:

In [ ]:
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
b = [2, 3, 4, 5]
b

[2, 3, 4, 5]

In [ ]:

a * b

array([[ 0,  3,  8, 15],
       [ 8, 15, 24, 35],
       [16, 27, 40, 55]])

В данном примере, без всяких циклов, мы умножили каждый столбец из массива `a` на соответствующий элемент из массива `b`.

Т.е. мы как бы транслировали (в какой-то степени можно сказать - растянули) массив `b` по массиву `a`.

То же самое мы можем проделать с каждой строкой массива `a`:

In [ ]:
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
c = [[10],
     [20],
     [30]]
a + c

array([[10, 11, 12, 13],
       [24, 25, 26, 27],
       [38, 39, 40, 41]])

In [ ]:
c

[[10], [20], [30]]

В данном случае мы просто прибавили к массиву `a` массив-столбец `c`. И получили, то что хотели.

При работе с двумерными или трехмерными массивами, особенно с массивами большей размерности, становится очень важным удобство работы с элементами массива, которые расположены вдоль отдельных измерений - его осей.

Например, у нас есть двумерный массив и мы хотим узнать его минимальные элементы по строкам и столбцам.

Для начала создадим массив из случайных чисел и пусть, для нашего удобства, эти числа будут целыми:

In [ ]:
a = np.random.randint(0, 15, size=(4, 6)) # (4, 6) - shape, структура массива
a

array([[ 8,  5,  2,  2,  1, 14],
       [13,  3,  1,  2,  5,  0],
       [13,  9,  9,  3, 11,  5],
       [11,  2,  6,  0,  5,  8]])

In [ ]:
test_gen_rnd = np.random.randint(150, 500, size=(5,5))
test_gen_rnd

array([[407, 491, 153, 166, 426],
       [498, 211, 488, 300, 343],
       [423, 200, 231, 409, 330],
       [350, 348, 443, 489, 159],
       [396, 243, 272, 363, 221]])

Минимальный элемент в данном массиве это:

In [ ]:
test_gen_rnd.min()

np.int64(153)

In [ ]:
# У списка так

min(lst)

11

In [ ]:
mass.min()

np.int32(11)

А вот минимальные элементы по столбцам и строкам:

In [ ]:
test_gen_rnd.min(axis=0)    #  минимальные элементы по столбцам

array([170, 171, 165, 196, 225], dtype=int32)

In [ ]:
test_gen_rnd.min(axis=1)    #  минимальные элементы по строкам

array([234, 314, 226, 170, 165], dtype=int32)

Такое поведение заложено практически во все функции и методы NumPy:

In [ ]:
test_gen_rnd.mean(axis=0)    #  среднее по столбцам

array([267.2, 341. , 266. , 291.2, 279.8])

In [ ]:
np.std(test_gen_rnd, axis=1)    #  стандартное отклонение по строкам

array([ 35.65389179,  50.13342199,  60.9832764 ,  58.39041017,
       102.686708  ])

Что насчет вычислений, их скорости и занимаемой памяти?

Для примера, создадим трехмерный массив:

In [ ]:
a = np.arange(48).reshape(4, 3, 4)
a

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]],

       [[24, 25, 26, 27],
        [28, 29, 30, 31],
        [32, 33, 34, 35]],

       [[36, 37, 38, 39],
        [40, 41, 42, 43],
        [44, 45, 46, 47]]])

Почему именно трехмерный?

На самом деле реальный мир вовсе не ограничивается таблицами, векторами и матрицами.

Еще существуют тензоры, кватернионы, октавы. А некоторые данные, гораздо удобнее представлять именно в трехмерном и четырехмерном представлении:

![image](https://raw.githubusercontent.com/dm-fedorov/numpy_basic/master/pic/3nd_array.jpg)

Визуализация (и хорошее воображение) позволяет сразу догадаться, как устроена индексация трехмерных массивов. Например, если нам нужно вытащить из данного массива число 31, то достаточно выполнить:

In [ ]:
a[2][1][3]    #  или a[2, 1, 3]

np.int64(31)

В самом деле, у массивов есть целый ряд важных атрибутов. Например, количество осей массива (его размерность), которую при работе с очень большими массивами, не всегда легко увидеть:

In [ ]:
a.ndim # размерность пространства

3

In [ ]:
mass.ndim # размерность пространства

1

In [ ]:
test_gen_rnd.ndim

2

Массив a действительно трехмерный.

Но иногда становится интересно, а на сколько же большой массив перед нами. Например, какой он формы, т.е. сколько элементов расположено вдоль каждой оси?
Ответить позволяет метод `ndarray.shape`:

In [ ]:
a.shape

(4, 3, 4)

Метод `ndarray.size` просто возвращает общее количество элементов массива:

In [ ]:
a.size

48

Еще может встать такой вопрос - сколько памяти занимает наш массив?

Иногда даже возникает такой вопрос - влезет ли результирующий массив после всех вычислений в оперативную память?

Что бы на него ответить надо знать, сколько "весит" один элемент массива:

In [ ]:
a.itemsize    #  эквивалентно ndarray.dtype.itemsize

8

`ndarray.itemsize` возвращает размер элемента в байтах.

Теперь мы можем узнать сколько "весит" наш массив:

In [ ]:
a.size * a.itemsize # сколько весит в памяти наш массив

384

Итого - 384 байта. На самом деле, размер занимаемой массивом памяти, зависит не только от количества элементов в нем, но и от испльзуемого типа данных:

In [ ]:
a.dtype

dtype('int64')

`dtype('int64')` - означает, что используется целочисленный тип данных, в котором для хранения одного числа выделяется 64 бита памяти.

Но если мы выполним какие-нибудь вычисления с массивом, то тип данных может измениться:

In [ ]:
b = a/3.14
b

array([[[ 0.        ,  0.31847134,  0.63694268,  0.95541401],
        [ 1.27388535,  1.59235669,  1.91082803,  2.22929936],
        [ 2.5477707 ,  2.86624204,  3.18471338,  3.50318471]],

       [[ 3.82165605,  4.14012739,  4.45859873,  4.77707006],
        [ 5.0955414 ,  5.41401274,  5.73248408,  6.05095541],
        [ 6.36942675,  6.68789809,  7.00636943,  7.32484076]],

       [[ 7.6433121 ,  7.96178344,  8.28025478,  8.59872611],
        [ 8.91719745,  9.23566879,  9.55414013,  9.87261146],
        [10.1910828 , 10.50955414, 10.82802548, 11.14649682]],

       [[11.46496815, 11.78343949, 12.10191083, 12.42038217],
        [12.7388535 , 13.05732484, 13.37579618, 13.69426752],
        [14.01273885, 14.33121019, 14.64968153, 14.96815287]]])

In [ ]:
b.dtype

dtype('float64')

Теперь у нас есть еще один массив - массив `b` и его тип данных `'float64'` - вещественные числа (числа с плавающей точкой) длинной 64 бита.

А его размер:

In [ ]:
b.size * b.itemsize

384

Скалярное произведение матриц

In [ ]:
a = np.array([[1, 2, 3, 4.0],
[5, 6, 7, 8],
[9, 10, 11, 12],
[3, 4, 5, 6]])
b = np.eye(4)
print(np.dot(a, b))
print(a.dot(b))
print(a @ b)

[[ 1.  2.  3.  4.]
 [ 5.  6.  7.  8.]
 [ 9. 10. 11. 12.]
 [ 3.  4.  5.  6.]]
[[ 1.  2.  3.  4.]
 [ 5.  6.  7.  8.]
 [ 9. 10. 11. 12.]
 [ 3.  4.  5.  6.]]
[[ 1.  2.  3.  4.]
 [ 5.  6.  7.  8.]
 [ 9. 10. 11. 12.]
 [ 3.  4.  5.  6.]]


преумножение всех элементов в массиве


In [ ]:
a = np.array([[1, 2, 3, 4.0],
[5, 6, 7, 8],
[9, 10, 11, 12],
[3, 4, 5, 6]])
print(np.prod(a))


172440576000.0


In [ ]:
a = np.array([[1, 20, 0, 6],
[5, 4, 7, 8.0],
[9, 0, 110,0],
[2, 4, 5, 6]])
print(np.sin(a))
print(np.cos(a))
print(np.tan(a))
print(np.sin(45))
print(np.cos(90))
print(np.tan(45))

[[ 0.84147098  0.91294525  0.         -0.2794155 ]
 [-0.95892427 -0.7568025   0.6569866   0.98935825]
 [ 0.41211849  0.         -0.04424268  0.        ]
 [ 0.90929743 -0.7568025  -0.95892427 -0.2794155 ]]
[[ 0.54030231  0.40808206  1.          0.96017029]
 [ 0.28366219 -0.65364362  0.75390225 -0.14550003]
 [-0.91113026  1.         -0.99902081  1.        ]
 [-0.41614684 -0.65364362  0.28366219  0.96017029]]
[[ 1.55740772  2.23716094  0.         -0.29100619]
 [-3.38051501  1.15782128  0.87144798 -6.79971146]
 [-0.45231566  0.          0.04428604  0.        ]
 [-2.18503986  1.15782128 -3.38051501 -0.29100619]]
0.8509035245341184
-0.4480736161291701
1.6197751905438615


In [ ]:
import numpy as np
a = np.array([[1, 20, 0, 6],
[5, 4, 7, 8.0],
[9, 0, 110,0],
[2, 4, 5, 6]])
print(np.sqrt(a))
print(np.sqrt(25))

[[ 1.          4.47213595  0.          2.44948974]
 [ 2.23606798  2.          2.64575131  2.82842712]
 [ 3.          0.         10.48808848  0.        ]
 [ 1.41421356  2.          2.23606798  2.44948974]]
5.0


Очень часто возникает задача создания массива определенного размера, причем, чем заполнен массив абсолютно неважно.

В этом случае можно воспользоваться циклами или генераторами списков (кортежей), но NumPy для таких случаев предлагает более быстрые и менее затратные функции-заполнители.

Функция `zeros` заполняет массив нулями, функция `ones` - единицами, а функция `empty` - случайными числами, которые зависят от состояния памяти.

По умолчанию, тип создаваемого массива - `float64`.

In [ ]:
np.zeros((3,3))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [ ]:
import numpy as np
a = np.zeros((2, 4), int)
print(a)

[[0 0 0 0]
 [0 0 0 0]]


In [ ]:
np.ones((3,3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [ ]:
import numpy as np
a = np.ones((2, 3))
print(a)

[[1. 1. 1.]
 [1. 1. 1.]]


In [ ]:
a = np.array([[1, 2, 3, 4.0],
[5, 6, 7, 8],
[9, 10, 11, 12]])
b = np.copy(a)
print(b)


[[ 1.  2.  3.  4.]
 [ 5.  6.  7.  8.]
 [ 9. 10. 11. 12.]]


In [ ]:
gen_X = np.arange(12, 100, 5)
print(gen_X)

[12 17 22 27 32 37 42 47 52 57 62 67 72 77 82 87 92 97]


In [ ]:
a = np.eye(4)
print(a)

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


**Линейная алгебра**

Довольно часто приходится вычислять ранг матриц:

In [ ]:
a = np.arange(1, 10).reshape(3, 3)
a

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [ ]:
np.linalg.matrix_rank(a)

2

In [ ]:
b = np.arange(1, 24, 2).reshape(3, 4)
b

array([[ 1,  3,  5,  7],
       [ 9, 11, 13, 15],
       [17, 19, 21, 23]])

In [ ]:
np.linalg.matrix_rank(b)

2

Еще чаще приходится вычислять определитель матриц, хотя результат вас может немного удивить:

In [ ]:
a = np.array([[1, 3], [4, 3]])
a

array([[1, 3],
       [4, 3]])

In [ ]:
np.linalg.det(a)

-8.999999999999998

In [ ]:
1*3 - 3*4    #  Результат должен быть целым числом

-9

В данном случае, из-за двоичной арифметики, результат нецелое число и округлять до ближайшего целого придется вручную.

Это связано с тем, что алгоритм вычисления определителя использует [LU-разложение](https://ru.wikipedia.org/wiki/LU-%D1%80%D0%B0%D0%B7%D0%BB%D0%BE%D0%B6%D0%B5%D0%BD%D0%B8%D0%B5) - это намного быстрее, чем обычный алгоритм, но за скорость все же приходится немного заплатить ручным округлением (конечно, если таковое требуется):

In [ ]:
np.linalg.det(a)

-8.999999999999998

In [ ]:
round(np.linalg.det(a))

-9.0

In [ ]:
b = np.arange(1, 48, 3).reshape(4, 4)
b

array([[ 1,  4,  7, 10],
       [13, 16, 19, 22],
       [25, 28, 31, 34],
       [37, 40, 43, 46]])

In [ ]:
np.linalg.det(b)

0.0

In [ ]:
round(np.linalg.det(b))

0

Транспонирование матриц:

In [ ]:
a

array([[1, 3],
       [4, 3]])

In [ ]:
a.T

array([[1, 4],
       [3, 3]])

Вычисление обратных матриц:

In [ ]:
a

array([[1, 3],
       [4, 3]])

In [ ]:
b = np.linalg.inv(a)
b

array([[-0.33333333,  0.33333333],
       [ 0.44444444, -0.11111111]])

In [ ]:
np.dot(a, b)

array([[1., 0.],
       [0., 1.]])

Решение систем линейных уравнений:

In [ ]:
#  система из двух линейных уравнений:
#  1*x1 + 5*x2 = 11
#  2*x1 + 3*x2 = 8
a = np.array([[1, 5], [2, 3]])
b = np.array([11, 8])

In [ ]:
x = np.linalg.solve(a, b)
x

array([1., 2.])

In [ ]:
np.dot(a, x)

array([11.,  8.])

**Статистика**

In [ ]:
a = np.random.randint(200, size=(5, 5))
a

array([[133, 119, 174, 139, 137],
       [ 26,   2, 156, 162,  63],
       [153, 186, 171, 136, 115],
       [ 90, 158, 115,  97,  28],
       [133,  74,  21, 109,  44]])

In [ ]:
#  Процентили:
np.percentile(a, 25)

74.0

In [ ]:
np.percentile(a, 50)

119.0

In [ ]:
np.percentile(a, 75)

153.0

Средние значения элементов массива и их отклонения:

In [ ]:
a = np.random.randint(13, size=(5, 5))
a

array([[ 6,  7,  9,  7,  2],
       [ 3,  7,  1,  3,  3],
       [ 9,  5,  1, 12, 11],
       [11,  4,  5,  4, 10],
       [ 0,  6,  7,  8,  9]])

In [ ]:
np.median(a)    #  медиана элементов массива

9.0

In [ ]:
np.mean(a)    #  среднее арифметическое

6.0

In [ ]:
np.var(a)    #  дисперсия

11.04

In [ ]:
np.std(a)    #  стандартное отклонение

3.3226495451672298

Корреляционные коэфициенты и ковариационные матрицы величин:

In [ ]:
x = np.array([1, 4, 3, 7, 10, 8, 14, 21, 20, 23])
y = np.array([4, 1, 6, 9, 13, 11, 16, 19, 15, 22])
z = np.array([29, 22, 24, 20, 18, 14, 16, 11, 9, 10])

In [ ]:
#  Линейный коэфициент корреляции Пирсона
#  величин 'x' и 'y'

XY = np.stack((x, y), axis=1)
XY

array([[ 1,  4],
       [ 4,  1],
       [ 3,  6],
       [ 7,  9],
       [10, 13],
       [ 8, 11],
       [14, 16],
       [21, 19],
       [20, 15],
       [23, 22]])

In [ ]:
np.corrcoef(XY)

array([[ 1., -1.,  1.,  1.,  1.,  1.,  1., -1., -1., -1.],
       [-1.,  1., -1., -1., -1., -1., -1.,  1.,  1.,  1.],
       [ 1., -1.,  1.,  1.,  1.,  1.,  1., -1., -1., -1.],
       [ 1., -1.,  1.,  1.,  1.,  1.,  1., -1., -1., -1.],
       [ 1., -1.,  1.,  1.,  1.,  1.,  1., -1., -1., -1.],
       [ 1., -1.,  1.,  1.,  1.,  1.,  1., -1., -1., -1.],
       [ 1., -1.,  1.,  1.,  1.,  1.,  1., -1., -1., -1.],
       [-1.,  1., -1., -1., -1., -1., -1.,  1.,  1.,  1.],
       [-1.,  1., -1., -1., -1., -1., -1.,  1.,  1.,  1.],
       [-1.,  1., -1., -1., -1., -1., -1.,  1.,  1.,  1.]])

In [ ]:
#  Кросс-корреляции:
np.correlate(x, y)

array([1736])

In [ ]:
np.correlate(x, z)

array([1486])

In [ ]:
#  Ковариационные матрицы:
np.cov(XY)

array([[ 4.5, -4.5,  4.5,  3. ,  4.5,  4.5,  3. , -3. , -7.5, -1.5],
       [-4.5,  4.5, -4.5, -3. , -4.5, -4.5, -3. ,  3. ,  7.5,  1.5],
       [ 4.5, -4.5,  4.5,  3. ,  4.5,  4.5,  3. , -3. , -7.5, -1.5],
       [ 3. , -3. ,  3. ,  2. ,  3. ,  3. ,  2. , -2. , -5. , -1. ],
       [ 4.5, -4.5,  4.5,  3. ,  4.5,  4.5,  3. , -3. , -7.5, -1.5],
       [ 4.5, -4.5,  4.5,  3. ,  4.5,  4.5,  3. , -3. , -7.5, -1.5],
       [ 3. , -3. ,  3. ,  2. ,  3. ,  3. ,  2. , -2. , -5. , -1. ],
       [-3. ,  3. , -3. , -2. , -3. , -3. , -2. ,  2. ,  5. ,  1. ],
       [-7.5,  7.5, -7.5, -5. , -7.5, -7.5, -5. ,  5. , 12.5,  2.5],
       [-1.5,  1.5, -1.5, -1. , -1.5, -1.5, -1. ,  1. ,  2.5,  0.5]])

In [ ]:
np.cov(x)

array(63.65555556)

NumPy предоставляет [порядка 30 функций](https://pyprog.pro/random_sampling_functions/random_sampling_functions.html), позволяющих генерировать случайные числа с самыми разными вероятностными распределениями:

In [ ]:
np.random.beta(0.1, 0.6, size=5)    #  бета распределение

array([7.92713309e-03, 2.87234368e-10, 4.09388263e-06, 5.53163533e-02,
       3.60681696e-04])

In [ ]:
np.random.gamma(shape=0.8, scale=1.7, size=5)    # гамма распределение

array([0.04923457, 1.76728972, 0.29066241, 0.08721191, 0.31709391])

In [ ]:
np.random.pareto(3.5, size=5)    #  Паретто распределение

array([0.24347805, 0.01003316, 1.23638848, 0.14471213, 0.88225519])

In [ ]:
np.random.chisquare(2.2, size=5)     #  хи-квадрат распределение

array([2.56589767, 6.74122626, 0.92194733, 3.29904345, 3.01434815])

Вы так же имеете доступ к состоянию генератора случайных чисел, а так же можете управлять им:

In [ ]:
#np.random.get_state()    #  Вы можете узнать состояние генератора

In [ ]:
np.random.seed(123)    #  устанавливает состояние генератора

In [ ]:
np.random.rand(5)

array([0.69646919, 0.28613933, 0.22685145, 0.55131477, 0.71946897])

## Примеры

#### 1. Память массива

Как определить объем памяти любого массива Numpy? (★☆☆)

`подсказка: np.size, np.itemsize`

In [23]:
new_zeros_arr = np.zeros((50,50))
print("%d байт" % (new_zeros_arr.size * new_zeros_arr.itemsize))
test_matrx = np.arange(10**7)
print("%d байт" % (test_matrx.size * test_matrx.itemsize))

20000 байт
80000000 байт


#### 2. Как получить документацию по функции в numpy? (★☆☆)
`подсказка: np.info`

In [ ]:
np.info(np.add)

#### 5. Найти индексы ненулевых элементов в массиве, сгенерированном от 10 до 100 (★☆☆)
`подсказка: np.nonzero`

In [26]:
# подготовка 
base_array = np.arange(10, 100)
# Шаг 2: Определяем количество элементов для замены на 0
num_zeros = int(len(base_array) * 0.2)  # Заменяем 20% элементов

# Шаг 3: Выбираем случайные индексы для замены
random_indices = np.random.choice(len(base_array), num_zeros, replace=False)

# Шаг 4: Заменяем выбранные элементы на 0
result_array = base_array.copy()
result_array[random_indices] = 0
# Вывод результата
print(f"Массив с нулями: {result_array}")
array_nz_index = np.nonzero(result_array) # поиск индексов
print(f'Массив без 0: {array_nz_index}') #Эта функция возвращает индексы всех ненулевых элементов в result_array.


Массив с нулями: [10 11 12 13  0 15  0  0  0 19 20 21 22 23 24 25 26  0 28  0 30 31  0 33
 34 35 36 37 38 39 40  0  0 43  0  0 46 47 48 49  0 51  0 53  0 55  0 57
 58 59 60 61 62 63 64 65 66 67 68 69 70  0 72 73 74 75  0 77 78 79 80 81
 82 83 84 85 86 87 88 89 90  0 92 93 94 95 96 97 98 99]
Массив без 0: (array([ 0,  1,  2,  3,  5,  9, 10, 11, 12, 13, 14, 15, 16, 18, 20, 21, 23,
       24, 25, 26, 27, 28, 29, 30, 33, 36, 37, 38, 39, 41, 43, 45, 47, 48,
       49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 62, 63, 64, 65, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 82, 83, 84, 85,
       86, 87, 88, 89]),)


#### 7. Создайте двумерный массив с 1 на границе и 0 внутри (★☆☆)
`Подсказка: array[1:-1, 1:-1]`

In [27]:
result_arr = np.ones((10,10))
result_arr[1:-1,1:-1] = 0
print(f'требуемый массив:\n {result_arr}')

требуемый массив:
 [[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


#### 9. Как добавить границу (заполненную нулями) вокруг существующего массива? (★☆☆)
`подсказка: np.pad`

In [28]:
arr_ones = np.ones((5,5))
# описывает матрицу по границе 0ми
arr_ones_with_border = np.pad(arr_ones.copy(), pad_width=2, mode='constant', constant_values=0)
print(f'Требуемый массив версия 1 с pad:\n {arr_ones_with_border}')

Требуемый массив версия 1 с pad:
 [[0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 1. 1. 1. 1. 0. 0.]
 [0. 0. 1. 1. 1. 1. 1. 0. 0.]
 [0. 0. 1. 1. 1. 1. 1. 0. 0.]
 [0. 0. 1. 1. 1. 1. 1. 0. 0.]
 [0. 0. 1. 1. 1. 1. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0.]]


#### 10. Создайте матрицу размером 5х5 со значениями 1,2,3,4 чуть ниже диагонали (★☆☆)
`подсказка: np.diag`

In [29]:
test_num_arr = np.diag(1+np.arange(4),k=-1) # работа с диагональю
print(f'Требуемый массив:\n {test_num_arr}')

Требуемый массив:
 [[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]


#### 11. Нормализуйте случайную матрицу размером 5х5 (★☆☆)
`подсказка: (x - mean)/std`

In [31]:
arr = np.random.random((5,5)) * 10
print(f'Исходный массив:\n {arr}')
normal_arr = (arr - np.mean (arr)) / (np.std (arr))
print(f'Требуемый массив:\n {normal_arr}')

Исходный массив:
 [[9.93539958 8.65903922 8.67725532 9.9750952  0.67240914]
 [7.56203197 8.58269082 3.33593084 5.82077745 3.89007897]
 [0.39793179 7.19424468 5.10974574 8.56663231 6.59215128]
 [1.2379238  5.81884748 6.86383222 1.29754439 1.43884104]
 [3.76248422 0.41127598 9.14717262 3.86499302 9.74165083]]
Требуемый массив:
 [[ 1.35970116  0.96466239  0.97030035  1.37198712 -1.50723238]
 [ 0.62513415  0.94103225 -0.68286149  0.08620873 -0.51135037]
 [-1.59218425  0.51130247 -0.1338585   0.93606208  0.3249521 ]
 [-1.33220327  0.0856114   0.40903846 -1.31375045 -1.27001856]
 [-0.55084147 -1.58805416  1.11574168 -0.51911457  1.29973513]]


#### Нормализация изображения

In [38]:
import numpy as np

# Исходное изображение с пикселями в диапазоне 0-255
image = np.array([[0, 128, 255],
                  [64, 192, 32]], dtype=np.float32)

# Нормализация в диапазон [0, 1]
normalized_image = image / 255.0
print(f'Нормализация в диапазон [0, 1]:\n {normalized_image}')


normalized_image = (image - np.mean(image)) / np.std(image)
print(f'Нормализация в диапазон [-1, 1]:\n {normalized_image}')
print(np.std(normalized_image))  # Должно быть около 1

Нормализация в диапазон [0, 1]:
 [[0.        0.5019608 1.       ]
 [0.2509804 0.7529412 0.1254902]]
0.35204256
Нормализация в диапазон [-1, 1]:
 [[-1.2457645   0.18008812  1.5948012 ]
 [-0.53283817  0.89301443 -0.8893013 ]]
0.99999994


#### 12. Умножьте матрицу 5х3 на матрицу 3х2 (реальное матричное произведение) (★☆☆)
`подсказка: np.dot`

In [39]:
test_arr = np.dot(np.ones((5,3)), np.ones((3,2)))
print(f'Требуемый массив:\n {test_arr}')

# Альтернативное решение для Python 3.5 и выше
test_arr_02 = np.ones((5,3)) @ np.ones((3,2))
print(f'Требуемый массив:\n {test_arr_02}')

Требуемый массив:
 [[3. 3.]
 [3. 3.]
 [3. 3.]
 [3. 3.]
 [3. 3.]]
Требуемый массив:
 [[3. 3.]
 [3. 3.]
 [3. 3.]
 [3. 3.]
 [3. 3.]]


#### 1. Как получить все даты, соответствующие ноябрю 2024 года? (★★☆)
`подсказка: np.arange(dtype=datetime64['D'])`

In [40]:
MIN_DATE = '2024-11'
MAX_DATE = '2024-12'
dates_arr = np.arange(MIN_DATE, MAX_DATE, dtype='datetime64[D]')
print(f'Требуемый массив:\n {dates_arr}')

Требуемый массив:
 ['2024-11-01' '2024-11-02' '2024-11-03' '2024-11-04' '2024-11-05'
 '2024-11-06' '2024-11-07' '2024-11-08' '2024-11-09' '2024-11-10'
 '2024-11-11' '2024-11-12' '2024-11-13' '2024-11-14' '2024-11-15'
 '2024-11-16' '2024-11-17' '2024-11-18' '2024-11-19' '2024-11-20'
 '2024-11-21' '2024-11-22' '2024-11-23' '2024-11-24' '2024-11-25'
 '2024-11-26' '2024-11-27' '2024-11-28' '2024-11-29' '2024-11-30']


#### 2. Как вычислить ((A+B)*(-A/2)) без создания промежуточных копий массивов? (★★☆)
`подсказка: np.add(out=), np.negative(out=), np.multiply(out=), np.divide(out=)`

In [41]:
# Создаем массивы A и B
A = np.ones(3)*1
B = np.ones(3)*2

# Функции NumPy с параметром out= выполняют операции "на месте"
np.add(A,B,out=B) # сложение
np.divide(A,2,out=A) # деление
np.negative(A,out=A) # умножить на -1
np.multiply(A,B,out=A)

array([-1.5, -1.5, -1.5])

#### 3. Извлеките целую часть из случайного массива положительных чисел, используя 4 различных метода (★★☆)
`hint: %, np.floor, astype, np.trunc`

In [46]:
random_arr = np.random.uniform(0, 100, size=10)
print(random_arr)
# Метод 1: Использование оператора %
print(f"\nМетод 1 (оператор %):{random_arr - random_arr%1}")
print(f"\nМетод 2 (оператор //):{random_arr // 1}")
print(f"\nМетод 3 (оператор floor):{np.floor(random_arr)}")
print(f"\nМетод 4 (оператор astype):{random_arr.astype(int)}")
print(f"\nМетод 5 (оператор trunc):{np.trunc(random_arr)}")

[64.93225964 84.68972563 14.65098539 39.29267908 53.77017863 42.96156625
 93.98885512 89.85631846 55.02209097 69.59054496]

Метод 1 (оператор %):[64. 84. 14. 39. 53. 42. 93. 89. 55. 69.]

Метод 2 (оператор //):[64. 84. 14. 39. 53. 42. 93. 89. 55. 69.]

Метод 3 (оператор floor):[64. 84. 14. 39. 53. 42. 93. 89. 55. 69.]

Метод 4 (оператор astype):[64 84 14 39 53 42 93 89 55 69]

Метод 5 (оператор trunc):[64. 84. 14. 39. 53. 42. 93. 89. 55. 69.]
